File: QtCompressedHelpExtractor/MainModule.vb
Imports System.Data.SQLite
Imports System.IO
Imports System.IO.Compression
Imports System.Linq

Module MainModule

    Private Const BUFFER_SIZE As Integer = 4095

    Private _contentFilePath As String
    Private _inputFilePath As String
    Private _outputPath As String
    Private _wkhtmltopdfPath As String

    Function Main() As Integer

        Directory.SetCurrentDirectory(AppDomain.CurrentDomain.BaseDirectory)

        If Environment.GetCommandLineArgs.Length < 3 Then
            Console.WriteLine("{0} <sqlite_file_path> <output_directory>", Process.GetCurrentProcess.ProcessName)
            Return 1
        End If

        _inputFilePath = Environment.GetCommandLineArgs(1)
        _outputPath = Environment.GetCommandLineArgs(2)

        If String.IsNullOrEmpty(_inputFilePath) OrElse Not File.Exists(_inputFilePath) Then
            Console.WriteLine("Error: SQLite file does not exist.")
            Console.ReadKey()
            Return 2
        End If

        Try
            ' Check if target path is valid.
            With New DirectoryInfo(_outputPath)
            End With

        Catch ex As Exception
            Console.WriteLine("Error: Incorrect target folder.")
            Console.ReadKey()
            Return 2
        End Try

        Try
            Console.WriteLine("Starting extraction ...")
            _contentFilePath = Path.Combine(_outputPath, "content.txt")

            ExportFileOrder()
            ExtractFilesFromSqlite()

            Console.WriteLine("Finished.")
            Return 0

        Catch ex As Exception
            Console.WriteLine("Unknown error: {0}{1}{2}", ex.Message, vbCrLf, ex.StackTrace)
            Return 1
        End Try
    End Function

    Private Sub ExportFileOrder()
        Dim data() As Byte
        Dim filePaths As IDictionary(Of String, String)
        Dim joinedPaths As String
        Dim pathList As IList(Of String)

        data = GetTableOfContentAsBlob()
        filePaths = ExtractFilePathsFromTableOfContent(data)
        pathList = filePaths.Keys.ToList()
        pathList = RemoveAnchorLinks(pathList)
        joinedPaths = String.Join(" ", pathList)

        Using writer As New StreamWriter(_contentFilePath)
            writer.WriteLine(joinedPaths)
        End Using
    End Sub

    Private Function RemoveAnchorLinks(filePaths As IList(Of String)) As IList(Of String)
        Dim withoutAnchor As String
        Dim withoutAnchors As New List(Of String)
        Dim regex As New Text.RegularExpressions.Regex("#{1}\w+")

        For Each path As String In filePaths
            withoutAnchor = regex.Replace(path, String.Empty)
            withoutAnchors.Add(withoutAnchor)
        Next

        Return withoutAnchors
    End Function

    Private Function ExtractFilePathsFromTableOfContent(data() As Byte) As Dictionary(Of String, String)
        Dim filePaths As New Dictionary(Of String, String)
        Dim filePath As String
        Dim filePathLength As Integer
        Dim i As Integer
        Dim lastIndex As Integer
        Dim level As Integer
        Dim title As String
        Dim titleLength As Integer

        lastIndex = data.Length - 1
        i = 0
        While i <= lastIndex
            i += 2 ' skip null-byte
            level = data(i + 1) + 256 * data(i) ' big endian
            i += 2

            If data(i) = &HFF AndAlso data(i + 1) = &HFF Then
                ' chapter has no file...
                filePath = String.Empty
                i += 4 ' consume four bytes
            Else
                ' chapter has file...
                i += 2 ' skip null-byte

                filePathLength = data(i + 1) + 256 * data(i) ' big endian
                i += 2

                filePath = Text.Encoding.BigEndianUnicode.GetString(data, i, filePathLength)
                i += filePathLength
            End If

            i += 2 ' skip null-byte

            titleLength = data(i + 1) + 256 * data(i) ' big endian
            i += 2

            title = Text.Encoding.BigEndianUnicode.GetString(data, i, titleLength)
            i += titleLength

            filePaths(filePath) = title
        End While

        Return filePaths
    End Function

    Private Function GetTableOfContentAsBlob() As Byte()
        Dim data As Byte() = Nothing

        Using dbConn As New SQLite.SQLiteConnection(String.Format("DataSource=""{0}""", _inputFilePath))
            dbConn.Open()

            Dim filePaths As New Dictionary(Of String, String)
            Dim totalCount As Integer = 0

            Using command As IDbCommand = dbConn.CreateCommand()
                command.CommandText =
                    "SELECT Data from ContentsTable " &
                    "WHERE ID = 1"

                Using reader As IDataReader = command.ExecuteReader
                    If reader.Read Then
                        data = reader.GetValue(0)
                    End If
                End Using
            End Using
        End Using

        Return data
    End Function

    Private Sub ExtractFilesFromSqlite()
        Using dbConn As New SQLiteConnection(String.Format("DataSource=""{0}""", _inputFilePath))
            dbConn.Open()

            Dim fileCount As Integer
            fileCount = GetFileCountFromSqlite(dbConn)

            Using command As IDbCommand = dbConn.CreateCommand()
                command.CommandText =
                    "SELECT Name, Data from FileDataTable " &
                    "JOIN FileNameTable ON FileDataTable.Id = FileNameTable.FileId " &
                    "WHERE Data NOT NULL AND Name NOT NULL " &
                    "ORDER BY FolderID, FileId"

                Dim count As Integer = 0

                Using reader As IDataReader = command.ExecuteReader
                    While reader.Read
                        Dim compressedData() As Byte
                        Dim data As Byte()
                        Dim name As String

                        name = reader.GetString(0)
                        compressedData = reader.GetValue(1)
                        data = DecompressData(compressedData)

                        WriteToFile(name, data)

                        count += 1
                        Console.WriteLine("{0} of {1} files done ({2}%).", count, fileCount, Math.Truncate(count * 100.0 / fileCount))
                    End While
                End Using
            End Using
        End Using
    End Sub

    Private Function GetFileCountFromSqlite(dbConn As SQLiteConnection) As Integer
        Dim fileCount As Integer = 0

        Using command As IDbCommand = dbConn.CreateCommand()
            command.CommandText =
                "SELECT COUNT(*) FROM FileNameTable " &
                "WHERE Name NOT NULL "

            Using reader As IDataReader = command.ExecuteReader
                If reader.Read Then
                    fileCount = reader.GetInt32(0)
                End If
            End Using
        End Using

        Return fileCount
    End Function

    Private Sub WriteToFile(name As String, data() As Byte)
        Dim absolutePath As String
        Dim fileName As String = Path.GetFileName(name)
        Dim directory As String = Path.GetDirectoryName(name)

        If String.IsNullOrEmpty(directory) Then
            directory = _outputPath
        Else
            directory = Path.Combine(_outputPath, directory)
        End If

        If Not IO.Directory.Exists(directory) Then
            IO.Directory.CreateDirectory(directory)
        End If

        absolutePath = Path.Combine(directory, fileName)

        Using writer As New FileStream(absolutePath, FileMode.Create)
            writer.Write(data, 0, data.Length)
        End Using
    End Sub

    Private Function AddGzipHeader(input() As Byte) As Byte()

        Using mem As New MemoryStream()
            mem.Write(New Byte() {&H1F, &H8B}, 0, 2)    ' GZip Magic Number
            mem.WriteByte(&H8)                          ' Compression Mode: Deflate
            mem.WriteByte(&H0)                          ' Flags: FTEXT
            mem.Write(input, 0, input.Length)           ' add input

            Return mem.ToArray
        End Using
    End Function

    Private Function DecompressData(compressed() As Byte) As Byte()
        Dim gzipData As Byte()
        Dim decompressed() As Byte

        gzipData = AddGzipHeader(compressed)
        decompressed = DecompressGzip(gzipData)

        Return decompressed
    End Function

    Private Function DecompressGzip(input As Byte()) As Byte()
        Using source = New MemoryStream(input)
            Dim buffer(BUFFER_SIZE) As Byte
            Dim read As Integer

            Using result As New MemoryStream()
                Using decompressionStream = New GZipStream(source, CompressionMode.Decompress)
                    Do
                        read = decompressionStream.Read(buffer, 0, buffer.Length)

                        If read > 0 Then
                            result.Write(buffer, 0, read)
                        End If
                    Loop While read > 0
                End Using

                Return result.ToArray
            End Using
        End Using
    End Function

End Module

File: QtCompressedHelpExtractor/MainModule.cs
﻿using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using System.Diagnostics;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text;
using Microsoft.Win32;

namespace QtCompressedHelpExtractor
{
    static class MainModule
    {
        private static string _contentFilePath;
        private static string _coverFile;
        private static string _inputFilePath;
        private static string _outputPath;
        private static string _outputPdfPath;
        private static string _wkhtmltopdf;

        public static int Main()
        {
            Directory.SetCurrentDirectory(AppDomain.CurrentDomain.BaseDirectory);
            if (Environment.GetCommandLineArgs().Length < 3)
            {
                Console.WriteLine("{0} <sqlite_file_path> <output_directory>", Process.GetCurrentProcess().ProcessName);
                return 1;
            }

            _coverFile = "wincc_oa.htm";
            _inputFilePath = Environment.GetCommandLineArgs()[1];
            _outputPath = Environment.GetCommandLineArgs()[2];
            _outputPdfPath = "output.pdf";
            _contentFilePath = Path.Combine(_outputPath, "content.txt");

            Directory.SetCurrentDirectory(_outputPath);

            if (string.IsNullOrEmpty(_inputFilePath) || !File.Exists(_inputFilePath))
            {
                Console.WriteLine("Error: SQLite file does not exist.");
                Console.ReadKey();
                return 2;
            }

            try
            {
                // Check if target path is valid.
                new DirectoryInfo(_outputPath);
            }
            catch
            {
                Console.WriteLine("Error: Incorrect target folder.");
                Console.ReadKey();
                return 2;
            }

            try
            {
                Console.WriteLine("Looking for wkhtmltopdf.");
                GetWkhtmltopdfPath();

                Console.WriteLine("Starting extraction ...");
                ExportFileOrder();
                ExtractFilesFromSqlite();

                Console.WriteLine("Creating the PDF: " + _outputPdfPath);
                CreatePdf();

                Console.WriteLine("Finished.");
                return 0;
            }
            catch (WkhtmltopdfNotFoundException)
            {
                Console.WriteLine("Wkhtmltopdf not found. Please install it first.");
                return 1;
            }
            catch (Exception ex)
            {
                Console.WriteLine("Unknown error: {0}\n{1}", ex.Message, ex.StackTrace);
                return 1;
            }
        }

        private static void GetWkhtmltopdfPath()
        {
            string regPath = @"SOFTWARE\wkhtmltopdf";
            using var regKey = Registry.LocalMachine.OpenSubKey(regPath);

            if (regKey is null)
            {
                throw new WkhtmltopdfNotFoundException("wkhtmltopdf not found, abort.");
            }

            _wkhtmltopdf = (string)regKey.GetValue("PdfPath");
        }

        private static void CreatePdf()
        {
            var content = string.Empty;
            var output = string.Empty;
            var args = "--footer-center \"[page]/[topage]\" " +
                "--enable-local-file-access " +
                "--load-error-handling skip " +
                "--read-args-from-stdin " +
                "--image-dpi 300 " +
                "--image-quality 80 " +
                "cover {0}";

            args = string.Format(args, _coverFile);

            using (var reader = new StreamReader(_contentFilePath))
            {
                if (!reader.EndOfStream)
                {
                    content = reader.ReadLine();
                    content += string.Format(" {0}", _outputPdfPath);
                }
            }

            Process process = new Process { 
                StartInfo = { 
                    FileName = _wkhtmltopdf, 
                    Arguments = args, 
                    UseShellExecute = false,
                    RedirectStandardInput = true,
                    RedirectStandardOutput = true,
                    CreateNoWindow = true
                } 
            };
            Console.WriteLine("Starting wkhtmltopdf, please wait...");
            process.Start();
            process.StandardInput.WriteLine(content);
            output = process.StandardOutput.ReadToEnd();
            Console.WriteLine(output);
            process.WaitForExit();
        }

        private static void ExportFileOrder()
        {
            var data = GetTableOfContentAsBlob();
            var filePaths = ExtractFilePathsFromTableOfContent(data);
            var pathList = RemoveAnchorLinks(filePaths.Keys.ToList());
            var joinedPaths = string.Join(" ", pathList);

            using var writer = new StreamWriter(_contentFilePath);
            writer.WriteLine(joinedPaths);
        }

        private static IList<string> RemoveAnchorLinks(IList<string> filePaths)
        {
            var withoutAnchors = new List<string>();
            var regex = new System.Text.RegularExpressions.Regex(@"#{1}\w+");

            foreach (string path in filePaths)
            {
                var withoutAnchor = regex.Replace(path, string.Empty);
                withoutAnchors.Add(withoutAnchor);
            }

            return withoutAnchors;
        }

        private static Dictionary<string, string> ExtractFilePathsFromTableOfContent(byte[] data)
        {
            var filePaths = new Dictionary<string, string>();
            var lastIndex = data.Length - 1;
            var i = 0;

            while (i <= lastIndex)
            {
                int filePathLength;
                int titleLength;

                string filePath;
                string title;

                i += 2; // skip null-bytes
                // level = data[i + 1] + 256 * data[i]; // big endian
                i += 2;

                if (data[i] == 0xFF && data[i + 1] == 0xFF)
                {
                    // chapter has no file...
                    filePath = string.Empty;
                    i += 4; // consume four bytes
                }
                else
                {
                    // chapter has file...
                    i += 2; // skip null-bytes
                    filePathLength = data[i + 1] + 256 * data[i]; // big endian
                    i += 2;
                    filePath = Encoding.BigEndianUnicode.GetString(data, i, filePathLength);
                    i += filePathLength;
                }

                i += 2; // skip null-bytes
                titleLength = data[i + 1] + 256 * data[i]; // big endian
                i += 2;
                title = Encoding.BigEndianUnicode.GetString(data, i, titleLength);
                i += titleLength;

                filePaths[filePath] = title;
            }
            return filePaths;
        }

        private static byte[] GetTableOfContentAsBlob()
        {
            byte[] data = null;
            using (var dbConn = new SQLiteConnection(string.Format("DataSource=\"{0}\"", _inputFilePath)))
            {
                dbConn.Open();
                var filePaths = new Dictionary<string, string>();

                using var cmd = dbConn.CreateCommand();
                cmd.CommandText =
                    "SELECT Data from ContentsTable " +
                    "WHERE ID = 1";

                using var reader = cmd.ExecuteReader();
                if (reader.Read())
                {
                    data = (byte[])reader.GetValue(0);
                }
            }

            return data;
        }

        private static void ExtractFilesFromSqlite()
        {
            using (var dbConn = new SQLiteConnection(string.Format("DataSource=\"{0}\"", _inputFilePath)))
            {
                dbConn.Open();
                var fileCount = GetFileCountFromSqlite(dbConn);

                using IDbCommand cmd = dbConn.CreateCommand();
                cmd.CommandText =
                    "SELECT Name, Data from FileDataTable " +
                    "JOIN FileNameTable " +
                      "ON FileDataTable.Id = FileNameTable.FileId " +
                    "WHERE Data NOT NULL AND Name NOT NULL " +
                    "ORDER BY FolderID, FileId";

                using var reader = cmd.ExecuteReader();
                var count = 0;
                while (reader.Read())
                {
                    var name = reader.GetString(0);
                    var compressedData = (byte[])reader.GetValue(1);
                    var data = DecompressData(compressedData);
                    WriteToFile(name, data);
                    count += 1;
                    Console.WriteLine("{0} of {1} files done ({2}%).", count, fileCount, Math.Truncate(count * 100.0 / fileCount));
                }
            }
        }

        private static int GetFileCountFromSqlite(SQLiteConnection dbConn)
        {
            var fileCount = 0;
            using (IDbCommand cmd = dbConn.CreateCommand())
            {
                cmd.CommandText =
                    "SELECT COUNT(*) FROM FileNameTable " +
                    "WHERE Name NOT NULL ";

                using var reader = cmd.ExecuteReader();
                if (reader.Read())
                {
                    fileCount = reader.GetInt32(0);
                }
            }

            return fileCount;
        }

        private static void WriteToFile(string name, byte[] data)
        {
            var fileName = Path.GetFileName(name);
            var directory = Path.GetDirectoryName(name);

            if (string.IsNullOrEmpty(directory))
            {
                directory = _outputPath;
            }
            else
            {
                directory = Path.Combine(_outputPath, directory);
            }

            if (!Directory.Exists(directory))
            {
                Directory.CreateDirectory(directory);
            }

            var absolutePath = Path.Combine(directory, fileName);
            using (var writer = new FileStream(absolutePath, FileMode.Create))
            {
                writer.Write(data, 0, data.Length);
            }
        }

        private static byte[] DecompressData(byte[] input)
        {
            input = input.Skip(6).ToArray();

            using (var sOutput = new MemoryStream())
            using (var sCompressed = new MemoryStream())
            {
                sCompressed.Write(input, 0, input.Length);
                sCompressed.Position = 0;
                using (var decomp = new DeflateStream(sCompressed, CompressionMode.Decompress))
                {
                    decomp.CopyTo(sOutput);
                }
                return sOutput.ToArray();
            }
        }
    }

    [Serializable]
    class WkhtmltopdfNotFoundException : Exception
    {
        public WkhtmltopdfNotFoundException()
        {

        }

        public WkhtmltopdfNotFoundException(string message)
        : base(message)
        {
        }

        public WkhtmltopdfNotFoundException(string message, Exception inner)
            : base(message, inner)
        {
        }
    }
}